src/mcts/node.cc

```diff
@@ -215,10 +232,67 @@ std::string Node::DebugString() const {
       << " WL:" << wl_ << " N:" << n_ << " N_:" << n_in_flight_
       << " Edges:" << static_cast<int>(num_edges_)
       << " Bounds:" << static_cast<int>(lower_bound_) - 2 << ","
-      << static_cast<int>(upper_bound_) - 2;
+      << static_cast<int>(upper_bound_) - 2
+      << " Solid:" << static_cast<bool>(solid_children_);
   return oss.str();
 }
 
+bool Node::MakeSolid() {
+  if (solid_children_ || num_edges_ == 0 || IsTerminal()) return false;
+  // Can only make solid if no immediate leaf childredn are in flight since we
+  // allow the search code to hold references to leaf nodes across locks.
+  Node* old_child_to_check = child_.get();
+  uint32_t total_in_flight = 0;
+  while (old_child_to_check != nullptr) {
+    if (old_child_to_check->GetN() <= 1 &&
+        old_child_to_check->GetNInFlight() > 0) {
+      return false;
+    }
+    if (old_child_to_check->IsTerminal() &&
+        old_child_to_check->GetNInFlight() > 0) {
+      return false;
+    }
+    total_in_flight += old_child_to_check->GetNInFlight();
+    old_child_to_check = old_child_to_check->sibling_.get();
+  }
+  // If the total of children in flight is not the same as self, then there are
+  // collisions against immediate children (which don't update the GetNInFlight
+  // of the leaf) and its not safe.
+  if (total_in_flight != GetNInFlight()) {
+    return false;
+  }
+  std::allocator<Node> alloc;
+  auto* new_children = alloc.allocate(num_edges_);
+  for (int i = 0; i < num_edges_; i++) {
+    new (&(new_children[i])) Node(this, i);
+  }
+  std::unique_ptr<Node> old_child = std::move(child_);
+  while (old_child) {
+    int index = old_child->index_;
+    new_children[index] = std::move(*old_child.get());
+    // This isn't needed, but it helps crash things faster if something has gone wrong.
+    old_child->parent_ = nullptr;
+    gNodeGc.AddToGcQueue(std::move(old_child));
+    Node* to_update_parent = new_children[index].child_.get();
+    if (!new_children[index].solid_children_) {
+      while (to_update_parent != nullptr) {
+        to_update_parent->parent_ = &new_children[index];
+        to_update_parent = to_update_parent->sibling_.get();
+      }
+    } else {
+      for (int i = 0; i < new_children[index].num_edges_; i++) {
+        to_update_parent[i].parent_ = &new_children[index];
```

Comment:
The name `to_update_parent` was a bit confusing initially, so a comment about fixing references to the solidified child could help or maybe name it `grandchild` ?

(If only there was some node iterator that could handle solid children or sibling chained 😜😜😜 )

Reply:
This code section does actually turn up twice, so I am going to refactor it and make it a bit nicer. It should be much clearer now.